### PR TITLE
support pub(restricted) in thread_local!

### PR DESCRIPTION
**Pull request by durka**
`pub(restricted)` was stabilized in #40556 so let's go!

Here is a [playground](https://play.rust-lang.org/?gist=f55f32f164a6ed18c219fec8f8293b98&version=nightly&backtrace=1).

I changed the interface of `__thread_local_inner!`, which is supposedly unstable but this is not checked for macros (#34097 cc @petrochenkov @jseyfried), so this may be an issue.

### CODE REVIEW

File: src/libstd/thread/local.rs [hunk 1]
```diff
@@ -130,31 +130,42 @@ impl<T: 'static> fmt::Debug for LocalKey<T> {
...
+    // handle multiple declarations where the first one is restricted public
+    ($(#[$attr:meta])* pub $vis:tt static $name:ident: $t:ty = $init:expr; $($rest:tt)*) => (
+        __thread_local_inner!($(#[$attr])* [pub $vis] $name, $t, $init);
+        thread_local!($($rest)*);
+    );
+    
+    // rule 4: handle multiple restricted public declarations
+    ($(#[$attr:meta])* pub $vis:tt static $name:ident: $t:ty = $init:expr) => (
+        __thread_local_inner!($(#[$attr])* [pub $vis] $name, $t, $init);
+        thread_local!($($rest)*);
```
Comment (durka) [inline src/libstd/thread/local.rs]: Please note these six rules can be reduced to four if we revive [the `$x:vis` RFC](https://github.com/rust-lang/rfcs/pull/1575).

**Comment by durka** [comment]
Github ate my comment, so just to reiterate: this PR is real, but also meant to demonstrate why we need [a `:vis` macro matcher](https://github.com/rust-lang/rfcs/pull/1575).

**Comment by alexcrichton** [comment]
Discussed at the libs triage meeting the other day, our conclusion was that we should probably wait for https://github.com/rust-lang/rust/pull/41012 to get merged, which looks like it's going to get merged.

**Comment by carols10cents** [comment]
Labeling waiting on review since rereview might be needed once #41012 is in.

**Comment by arielb1** [comment]
#41012 has landed. @alexcrichton can you look at this?

**Comment by durka** [comment]
I was planning to update or resubmit after updating the macro to use :vis.

**Comment by alexcrichton** [comment]
Ok thanks @durka!

**Comment by durka** [comment]
Actually we can't use `:vis` until #41012 hits beta, right?

**Comment by sfackler** [comment]
Seems like you can do it before then with `#[cfg(stage0)]` and `#[cfg(not(stage0))]` versions of the macro, right?

**Comment by durka** [comment]
Hmm, this is odd:

```
[00:17:04] Building stage1 compiler artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
... snip ...
[00:17:20]    Compiling syntax_pos v0.0.0 (file:///checkout/src/libsyntax_pos)
[00:17:20] error: :vis fragment specifier is experimental and subject to change (see issue #41022)
[00:17:20]  --> <thread_local macros>:2:29
[00:17:20]   |
[00:17:20] 2 | $ ( # [ $ attr : meta ] ) * $ vis : vis static $ name : ident : $ t : ty = $
[00:17:20]   |                             ^^^^^^^^^^^
[00:17:20]   |
[00:17:20]   = help: add #![feature(macro_vis_matcher)] to the crate attributes to enable
[00:17:20] 
[00:17:20] error: :vis fragment specifier is experimental and subject to change (see issue #41022)
[00:17:20]  --> <thread_local macros>:6:29
[00:17:20]   |
[00:17:20] 6 | $ ( # [ $ attr : meta ] ) * $ vis : vis static $ name : ident : $ t : ty = $
[00:17:20]   |                             ^^^^^^^^^^^
[00:17:20]   |
[00:17:20]   = help: add #![feature(macro_vis_matcher)] to the crate attributes to enable
```

But I did [add that attribute](https://github.com/rust-lang/rust/pull/40984/files#diff-242481015141f373dcb178e93cffa850R285). Does it have to be added to crates that invoke the macro as well? That would be a problem. And it should be fine because `thread_local!` and `__thread_local_inner!` are marked `#[allow_internal_unstable]`. Perhaps fragment specifiers are not covered by that?

**Comment by alexcrichton** [comment]
As the author of https://github.com/rust-lang/rust/pull/41012, you're probably in the best position to answer that question :)

**Comment by durka** [comment]
Good call :)

Can a commit fixing that go in this PR or should I separate it?

**Comment by alexcrichton** [comment]
Nah I think it's fine to be here as well. Two final pieces:

* Looks like a test may be failing on Travis?
* To clarify, this was possible *without* the `:vis` matcher, right? That is, if `:vis` isn't stabilized, we can still hope to implement this, right? (albeit not as nicely)

**Comment by durka** [comment]
Yes, we can definitely do it without `:vis`. The travis failure is connected to the fact that unfortunately, `:vis` is not as helpful as I thought it would be here. We have the pattern

```rust
($(#[$attr:meta])* $vis:vis ...)
```

and the macro parser won't let `:vis` terminate that repetition. I've reached to @jseyfried to see if this is fixable or not -- if not, the macro doesn't really get any shorter.

**Comment by jseyfried** [comment]
>  I've reached out to @jseyfried to see if this is fixable or not

I believe this is fixable, but I'm not sure exactly how -- I'll look into it when I get the time (probably week or two).

**Comment by durka** [comment]
OK, I did come up with a way to make it shorter, by having the macro munch attributes one at a time until it gets to the visibility specifier:

```rust
macro_rules! thread_local {
    // terminate recursion
    (@ [] -> $_x:tt) => {};
    
    // munch one attribute
    (@ [#[$attr:meta] $($rest:tt)*] -> [$(#[$attrs:meta])*]) => {
        thread_local!(@ [$($rest)*] -> [$(#[$attrs])* #[$attr]]);
    };
    
    // finish a static with no trailing semicolon (therefore there are no more)
    (@ [$vis:vis static $name:ident: $typ:ty = $init:expr] -> [$(#[$attrs:meta])*]) => {
        __thread_local_inner!(($(#[$attrs:meta])*) $vis, $name, $typ, $init);
    };
    
    // finish a static with trailing semicolon and continue to parse more
    (@ [$vis:vis static $name:ident: $typ:ty = $init:expr; $($rest:tt)*] -> [$(#[$attrs:meta])*]) => {
        __thread_local_inner!(($(#[$attrs:meta])*) $vis, $name, $typ, $init);
        thread_local!(@ [$($rest)*] -> []);
    };
    
    // public entry point
    ($($t:tt)*) => {
        thread_local!(@ [$($t)*] -> []);
    }
}
```

This business with the `@` "private" rules could be pulled into a separate macro as well (e.g. `__thread_local_collector!`). If only that trailing semicolon was required, we could remove one more rule, oh well.

**Comment by alexcrichton** [comment]
Yeah if we do add a macro implementation like that we'd want it to be an internal macro hidden away, but it seems reasonableto have such an implementation.

**Comment by alexcrichton** [comment]
@durka to clarify, did you intend on landing that instantiation in this PR? Or waiting for a fix to the macros from @jseyfried?

**Comment by durka** [comment]
Ah sorry this dropped off my radar. I'm happy to land either the version
with extra arms (so not actually using :vis in thread_local!, though it
would be used in thread_local_inner!) or the muncher version in this PR.
The muncher version wouldn't look great in rustdoc (since it would just be
$($t:tt)*) but we can solve that with docs. Do you care which way we go?

**Comment by alexcrichton** [comment]
I'd ideally prefer to wait for `:vis` bugs to get fixed (or at least see if they're not fixable) but I also don't have a burning need for this change just yet. If there's some strong motivation I could see us landing this sooner rather than later.

**Comment by carols10cents** [comment]
> I'd ideally prefer to wait for :vis bugs to get fixed (or at least see if they're not fixable)

Are there issues/PRs for the `:vis` bugs that are blocking this PR that we could link to for status tracking purposes?

**Comment by durka** [comment]
I think it is just #26444 (closed WONTFIX) unfortunately. But I could be wrong.

**Comment by alexcrichton** [comment]
This was to a large degree the motivation for https://github.com/rust-lang/rust/pull/41012, right? If so, it seems a shame if it's actually still difficult to leverage :(

**Comment by Mark-Simulacrum** [comment]
@durka Could you give an update here? Looks like Travis is still failing -- @jseyfried, did you have a chance to [take a look](https://github.com/rust-lang/rust/pull/40984#issuecomment-296796073)?

**Comment by durka** [comment]
Status quo.

**Comment by durka** [comment]
I'm going to close this for now and post in the `:vis` tracking issue. I'll resubmit once we decide on an approach, or if someone complains about `thread_local!` specifically.

**Comment by durka** [comment]
Rebased and reopened now that #42913 got merged.

**Comment by Mark-Simulacrum** [comment]
Looks like you'll need to open a new pull request -- GH is being annoying and not letting this one be reopened.

**Comment by durka** [comment]
Resubmitted as #43185.

**Pull request closed without merging**